### ISSUE TITLE
Make workload_x509_svid_key_type can be configured per SPIFFE entry

### ISSUE THREAD

**Issue by itomsawyer** [spiffe/spire#4483]
Hi all, 

Spire agent has a configuration option `workload_x509_svid_key_type` for choosing workload private key type using rsa-2048 or ec-p256. 

But it is a agent scope configuration, I wonder if there’s plan to make this configuration per entry just like `svid_x509_ttl` and `svid_jwt_ttl` . 

I think it make sense when any of workload is required to use specific key type where others don’t.

For now only two type of key can be generated by Spire agent , perhaps there will be more in the future.

If it make sense, I would like to  implement it.

Thx.

* **Version**: 
* **Platform**: 
* **Subsystem**:

**Comment by MarcosDY** [comment]
Initially, we analyzed having the key-type configurable per workload, but there was not a clear use case for that at that moment and one of the concerns was that allowing this could make it hard to follow SVIDs per agent, so we chose to keep it simple.

Do you have a use case where you need to define the key-type per workload?

**Comment by itomsawyer** [comment]
SPIRE agent/server key pair Generator is defined in 

https://github.com/spiffe/spire/blob/a6ce058b58408b1f2fa80ba772e447d592344a0f/pkg/agent/plugin/keymanager/base/keymanagerbase.go#L39

For now a defaultGenerator implements  Generator interface which use the native golang crypto library.

We have a case that several kind of workloads need offload the procedure to hardware on  node to generate other kinds of key pair instead of RSA/ECC. Meanwhile, most of workloads is ok to use RSA/ECC key pair with native golang.

Actually,  per SPIFFE entry workload_x509_svid_key_type configuration is a prerequisite to achieve above goal.

**Comment by azdagron** [comment]
The agent manages two kinds of keys:
1. The agent's key
   - backed by the KeyManager interface
   - only used internally by the agent
1. The workload keys
   - backed by memory
   - conveyed to Workloads over the Workload API

The code you linked only applies to the agent key. Supporting a hardware backed key for the agent key is exactly what the keymanager interface was designed for.

However, supporting hardware backed keys for Workloads presents many challenges.
- How would the Workload instead use a hardware backed key?
- How would the agent inform the Workload which hardware backed key to use?
- How would the agent know which hardware backed key to sign an SVID over for a given workload (it needs the public key at a minimum in order to get an SVID signed over the public key)?

**Comment by itomsawyer** [comment]
@azdagron Thanks for the clearly statements of how agent managing two kinds of keys.

I can tell key management plugin is designed for maintaining agent key. And what I mean to do is extending workload key pair generator to generate key pair.

> How would the Workload instead use a hardware backed key?
I think it is kind of a workload key type configuration. It tells agent to generate a SVID with backend hardware key pair

> How would the agent inform the Workload which hardware backed key to use?
 I think key pair type is per SVID entry configuration, maybe it can be distinguished by entry ID.

> How would the agent know which hardware backed key to sign an SVID over for a given workload (it needs the public key at a minimum in order to get an SVID signed over the public key)?
Hardware backend engine can only generate key pair and let SPIRE agent to store it.

**Comment by evan2645** [comment]
> I think it is kind of a workload key type configuration. It tells agent to generate a SVID with backend hardware key pair

The workload api interface requires us to give the private key to the workload .. how do you envision the workload interacting with this hardware-backed key? Will the agent need to e.g. perform signing operations on behalf of the workload?

**Comment by itomsawyer** [comment]
@evan2645  

![image](https://github.com/spiffe/spire/assets/1269823/c1c37f0b-cf26-4bdc-a245-86bbe8e5fd46)

For now , I think how SPIRE agent generate SVID based on register entry as above.  Agent generate key pair and CSR on behalf of workload. Then, agent request server to sign the CSR to get a X509 certificate. Agent store the private key and X509 certificate in SVID store. When workload request agent to get a valid SVID after workload attest, agent delivery the proper SVID to workload. 

If I made any mistake , please let me know.

So, I think there's two separated phases can be enhanced. 

1. Generate public/private key pair.
2. SVID store

For the first one, agent now use golang crypto lib to generate key pair . It can be done by hardware or other software library.  If we left SVID store unchanged , I think it has no effect of current procedure to delivery SVID(including private key) to workload.  When talking about hardware backend key , I'm not mean to indicate that key will never leave hardware.

Secondly, SVID store is memory only. When agent restart，all entries will regenerate their SVIDs. I wonder if it is possible to use other utility to store SVID persistently.

**Comment by evan2645** [comment]
Thanks @itomsawyer for the clarification, it's very useful ❤️ 

I have a follow-on question given your response ... is your goal to have hardware-backed key generation and management in order to achieve better confidentiality and/or more performant generation? Or is the goal to make SPIRE compatible with key types and hash algorithms that it doesn't currently support today?

**Comment by itomsawyer** [comment]
Thanks @evan2645  for the quick reply. 

My answer is both.

> is your goal to have hardware-backed key generation and management in order to achieve better confidentiality and/or more performant generation?

We are facing a production environment that need all key generation is certified by specific security standard. Some hardware is certified and need to be used.

> Or is the goal to make SPIRE compatible with key types and hash algorithms that it doesn't currently support today?

Yes, we need to use key types and hash algorithms as specified in https://datatracker.ietf.org/doc/html/rfc8998

**Comment by evan2645** [comment]
> Yes, we need to use key types and hash algorithms as specified in https://datatracker.ietf.org/doc/html/rfc8998

I think supporting this will be more difficult. The KeyManager interface knows about key types and algorithms. I think it also means different kinds of encoding inside X.509 certificates, and different key type inside JWT. So I don't think offloading to hardware alone is enough to solve it?

**Comment by itomsawyer** [comment]
@evan2645 

> So I don't think offloading to hardware alone is enough to solve it?

Definitely agree with you. So I make a propose to implement part of the whole works to achieving this step by step.

**Comment by evan2645** [comment]
Ok, this is helpful, thanks again for all the information @itomsawyer 

The scope of what's needed to support this is pretty large, and the work will likely span many releases. We should plan out the work so we know what needs to be released when, and how much work there is for each release. If we want to support it, I think the minimum is a new plugin type and a major refactor of the server's key manager interface. I'd say it's probably 6-8 months of on-again off-again work ... do you think it's something you'd be able to commit to contributing? I'm happy to continue exploring what this might look like if so

**Comment by itomsawyer** [comment]
@evan2645  Thanks for your agreements with the feature is make sense. I would like to create a new plugin which maybe named as WorkloadKeyManagement for generating private/public key pair . I think I can create a new issue for tracking it.

**Comment by evan2645** [comment]
> I would like to create a new plugin which maybe named as WorkloadKeyManagement for generating workload private/public key pair

Yes, I know this is likely to be part of the work involved, but there's a lot of other pieces of work that we'll also need in order for the feature to make sense for SPIRE as it is today. So what I'm trying to say is that we should have a good understanding first of what all of those changes are going to be, so we know and can agree upon the overall scope of the work, _before_ implementing any of the features.

I'm happy to create a new issue, though I would probably name it something like "SPIRE support for crypto agility" , and the conversation there would be about coming up with a design and set of supporting changes that we'll need to make in order to achieve the goal. The most important thing is that we get some commitment from one or more contributors to carry the work through the end (design, implementation, and roll-out). It will be bad for SPIRE to have the work only half-done. So the main question is whether you or any other contributors are willing to invest the time and effort to get this new capability shipped

**Comment by itomsawyer** [comment]
@evan2645  Thanks for your patient to finely explain the point of view.

I would like to contribute with the whole scope of work with "SPIRE support for crypto agility".  Maybe two contributors  of our team will invest effort to get the new capability done.

> I'm happy to create a new issue, though I would probably name it something like "SPIRE support for crypto agility" , and the conversation there would be about coming up with a design and set of supporting changes that we'll need to make in order to achieve the goal.

For the kick-off start,  I'm willing to take participate with the design work and list changes to be done. Maybe i don't have a good understanding of the whole design thought of Spire, if it is ok for you to make a scaffold of this proposal, and then I can  put my thought into it

**Comment by evan2645** [comment]
Hi @itomsawyer , thanks for your patience. I opened https://github.com/spiffe/spire/issues/4698 in support of this work, it would be great to have your thoughts there, and particularly some help with ideas on how to address each of the bullet points I left there

I'm gonna go ahead and close this out as we can move conversation to https://github.com/spiffe/spire/issues/4698 ... thanks for opening this 🙏
